test-js/unit/pentaho/CustomContextVars.Spec.js

```diff
@@ -0,0 +1,113 @@
+/*!
+ * Copyright 2010 - 2015 Pentaho Corporation.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+define([
+  "tests/test-utils"
+], function(testUtils) {
+
+  "use strict";
+
+  /* global describe:false, it:false, expect:false, beforeEach:false, afterEach:false */
+
+  // Use alternate, promise-aware version of `it`.
+  var it = testUtils.itAsync;
+
+  describe("pentaho.CustomContextVars -", function() {
+
+    /**
+     * Mocks AMD with dummy default context variables.
+     */
+    function configDefaultContext(localRequire) {
+
+      localRequire.define("pentaho/contextVars", {
+        application: "APP",
+        user:        "USER",
+        theme:       "THEME",
+        locale:      "LOCALE"
+      });
+    }
+
+    it("is a function", function() {
+
+      return require.using(["pentaho/CustomContextVars"], configDefaultContext, function(CustomContextVars) {
+        expect(typeof CustomContextVars).toBe("function");
+      });
+    });
+
+    describe("new CustomContextVars(spec) -", function() {
+
+      it("should return a CustomContextVars instance", function() {
+
+        return require.using(["pentaho/CustomContextVars"], configDefaultContext, function(CustomContextVars) {
+          var contextVars = new CustomContextVars();
+          expect(contextVars instanceof CustomContextVars).toBe(true);
+        });
+      });
+
+      /**
+       * Tests that a context variable is properly implemented by pentaho.CustomContextVars.
+       *
+       * @param {string} varName - The name of the context variable's.
+       */
+      function testVariable(varName) {
+        describe(varName, function() {
+
+          it("should default to the default variable's value, when spec." + varName + " is not specified", function() {
+
+            return require.using([
+              "pentaho/contextVars",
+              "pentaho/CustomContextVars"
+            ], configDefaultContext, function(defaultContextVars, CustomContextVars) {
+
+              var contextVars = new CustomContextVars();
+              expect(contextVars[varName]).toBe(defaultContextVars[varName]);
+            });
+          });
+
+          it("should default to the default variable's value, when spec." + varName + " is empty", function() {
```

Comment:
Usually, I exclude the empty strings as valid values and just interpret them as null/undefined. In most contexts, there's no meaningful use for an empty string.
But usually, also, I'd document these cases as `?nonEmptyString` and it was not the case as they are documented as `?string` (https://github.com/pentaho/pentaho-platform-plugin-common-ui/blob/master/package-res/resources/web/pentaho/_doc/spec/IContextVars.jsdoc). 
This type does not include `undefined` (would have to declare it as `string|null|undefined`).

Assuming only valid types (except `undefined`) are provided, the empty string — a member of the declared type — is converted to `null`. No attempt is made to handle values of other types (except `undefined`).

Are there any use cases for distinguishing between `null` and `undefined` (like context inheritance) ?
Do you want me to improve the code to make sure that any non-string values are converted to the default?